index.js

```diff
@@ -77,91 +173,57 @@ var agent;
 function start(projectConfig) {
   var config = initConfig(projectConfig);
 
-  if (traceApi.isActive() && !config.forceNewAgent_) { // already started.
+  if (traceAgent && !config.forceNewAgent_) { // already started.
     throw new Error('Cannot call start on an already started agent.');
+  } else if (traceAgent) {
+    // For unit tests only.
+    // Undoes initialization that occurred last time start() was called.
+    stop();
   }
 
   if (!config.enabled) {
-    return traceApi;
+    return traceAgent;
   }
 
-  if (config.logLevel < 0) {
-    config.logLevel = 0;
-  } else if (config.logLevel >= common.logger.LEVELS.length) {
-    config.logLevel = common.logger.LEVELS.length - 1;
-  }
   var logger = common.logger({
     level: common.logger.LEVELS[config.logLevel],
     tag: '@google-cloud/trace-agent'
   });
-
-  if (config.projectId) {
-    logger.info('Locally provided ProjectId: ' + config.projectId);
-  }
-
-  if (onUncaughtExceptionValues.indexOf(config.onUncaughtException) === -1) {
-    logger.error('The value of onUncaughtException should be one of ',
-      onUncaughtExceptionValues);
-    throw new Error('Invalid value for onUncaughtException configuration.');
-  }
-
-  var headers = {};
-  headers[constants.TRACE_AGENT_REQUEST_HEADER] = 1;
+  // CLS namespace for context propagation
+  cls.createNamespace();
+  TraceWriter.create(logger, config);
```

Comment:
Could we get the metadata and `then` create the TraceWriter, passing in all necessary information into the constructor and getting rid of `setMetadata`? Either way, it seems like TraceWriter is a singleton, and arbitrary calls to `TraceWriter.get()` can return either an initialized or uninitialized (`setMetadata` hasn't been called yet) TraceWriter. It would be better to return null, or a fully initialized writer.

I think it would be better for `TraceWriter.get